delft/sequenceLabelling/models.py

```diff
@@ -364,16 +360,76 @@ def __init__(self, config, ntags=None):
         self.model = Model(inputs=[word_input, char_input, casing_input, length_input], outputs=[pred])
         self.config = config
 
+class BidLSTM_CRF_FEATURES(BaseModel):
+    """
+    A Keras implementation of BidLSTM-CRF for sequence labelling using tokens combined with 
+    additional generic discrete features information.
+    """
+
+    name = 'BidLSTM_CRF_FEATURES'
+
+    def __init__(self, config, ntags=None):
+
+        # build input, directly feed with word embedding by the data generator
+        word_input = Input(shape=(None, config.word_embedding_size), name='word_input')
+
+        # build character based embedding
+        char_input = Input(shape=(None, config.max_char_length), dtype='int32', name='char_input')
+        char_embeddings = TimeDistributed(Embedding(input_dim=config.char_vocab_size,
+                                    output_dim=config.char_embedding_size,
+                                    mask_zero=True,
+                                    #embeddings_initializer=RandomUniform(minval=-0.5, maxval=0.5),
+                                    name='char_embeddings'
+                                    ))(char_input)
+
+        chars = TimeDistributed(Bidirectional(LSTM(config.num_char_lstm_units,
+                                                   return_sequences=False)))(char_embeddings)
+
+        # layout features input and embeddings
+        features_input = Input(shape=(None, len(config.features_indices)), dtype='float32', name='features_input')
+
+        # The input dimension is calculated by
+        # features_vocabulary_size (default 12) * number_of_features + 1 (the zero is reserved for masking / padding)
+        features_embedding = TimeDistributed(Embedding(input_dim=config.features_vocabulary_size * len(config.features_indices) + 1,
+                                       output_dim=config.features_embedding_size,
+                                       # mask_zero=True,
+                                       trainable=False,
```

Comment:
I am just looking at it again and comparing it to what I have.
I have a perhaps silly question (not sure if it's the right place but is related to the change):
This features `Embedding` layer is marked as not trainable. Which I think it means it will initialized with random values by default. Are you initializing it somewhere else? Or why is it not trainable?

Reply:
The input feature channel is initialized by the preprocessor. When it's not trainable normally we just have a `uniform` mapping of the one-hot encoding to features_embedding_size dimension (Glorot uniform distribution is the default if I am not wrong, as defined [here](https://keras.io/api/layers/initializers/)) - so it's not random in the sense that a small change in one of the feature should lead to a small shift in the target space. 

Our hope was that if we set it to `trainable=True`, we would better fit the training actual features and get better performance. However, the [experiments](https://grobid.readthedocs.io/en/latest/Benchmarking-models/#accuracy) have not shown improvement, so it might overfit or whatever.

Reply:
Thank you. I think I am still slightly not sure whether you confirmed that the embedding layer's weight is using weights initialized by Keras? I understand that the preprocessor is providing the inputs. But does it provide embedding weights as well?

Reply:
Yes sorry it's not clear! Yes the weights of the embedding layer are initialized randomly with a Glorot uniform distribution, and then the input of the embedding layer is the preprocessor one-hot encoding of the features. So the preprocessor does not provide the embedding weights, just the "index" (as one-hot encoded values) that will be mapped via the uniform distribution in the `output_dim` space.